python/anvill/binja.py

```diff
@@ -173,11 +173,64 @@ def _convert_bn_llil_type(
         ret = IntegerType(reg_size_bytes, True)
         return ret
 
+def _cache_key(tinfo: bn.types.Type) -> str:
+    """ Convert bn Type instance to cache key"""
+    return str(tinfo)
+
+def _convert_named_type_reference(
+    bv, tinfo: bn.types.Type, cache) -> Type:
+    """ Convert named type references into a `Type` instance"""
+    if tinfo.type_class != bn.TypeClass.NamedTypeReferenceClass:
+        return
+
+    named_tinfo = tinfo.named_type_reference
+    if (named_tinfo.type_class
+        == bn.NamedTypeReferenceClass.StructNamedTypeClass):
+        # Get the bn struct type and recursively recover the elements
+        ref_type = bv.get_type_by_name(named_tinfo.name);
+        struct_type =  ref_type.structure
+        ret = StructureType()
+        cache[_cache_key(struct_type)] = ret
+        for elem in struct_type.members:
+            ret.add_element_type(_convert_bn_type(bv, elem.type, cache))
+        return ret
+
+    elif (named_tinfo.type_class
+        == bn.NamedTypeReferenceClass.UnionNamedTypeClass):
+        # Get the union type and recover the member elements
+        ref_type = bv.get_type_by_name(named_tinfo.name);
+        struct_type =  ref_type.structure
+        ret = UnionType()
+        cache[_cache_key(struct_type)] = ret
+        for elem in struct_type.union.members:
+            ret.add_element_type(_convert_bn_type(bv, elem.type, cache))
+        return ret
+
+    elif (named_tinfo.type_class
+        == bn.NamedTypeReferenceClass.TypedefNamedTypeClass):
+        ref_type = bv.get_type_by_name(named_tinfo.name);
+        ret = TypedefType()
+        cache[_cache_key(ref_type)] = ret
+        ret.set_underlying_type(_convert_bn_type(bv, ref_type, cache))
+        return ret
 
-def _convert_bn_type(tinfo: bn.types.Type, cache):
+    elif (named_tinfo.type_class
+        == bn.NamedTypeReferenceClass.EnumNamedTypeClass):
+        # Set the underlying type int of size width
+        ref_type = bv.get_type_by_name(named_tinfo.name);
+        ret = EnumType()
+        cache[_cache_key(ref_type)] = ret
+        ret.set_underlying_type(IntegerType(ref_type.width, False))
+        return ret
+
+    else:
+        DEBUG("WARNING: Unknown named type {} not handled".format(named_tinfo))
+
+
+def _convert_bn_type(bv, tinfo: bn.types.Type, cache):
     """Convert an bn `Type` instance into a `Type` instance."""
-    if str(tinfo) in cache:
-        return cache[str(tinfo)]
+    if _cache_key(tinfo) in cache:
```

Comment:
Save `_cache_key(tinfo)` to a variable and re-use it everywhere in this function.